### PR TITLE
Threatzone Integration

### PR DESCRIPTION
**Pull request by vvelitkn-malwation**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Malwation Threatzone Integration

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+
+
+def threatzone_check_limits(client: Client) -> CommandResults:
+    availability = client.threatzone_check_limits(None)
+    readable_output = tableToMarkdown('LIMITS', availability["Limits"])
+    return CommandResults(
+        readable_output=readable_output,
+        outputs_key_field='results',
+        outputs={'ThreatZone.Limits(val.Job_ID == obj.Job_ID)': availability["Limits"]}
+    )
```
Comment (ilappe) [inline Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py:366]: The `ComandREsults` object has the `outputs_prefix` and `outputs_key_field` properties which together combined the old key `ThreatZone.Limits(val.Job_ID == obj.Job_ID)` 
please use this syntax in all the CommandaResults objects it's our new convention 
see https://xsoar.pan.dev/docs/integrations/code-conventions#commandresults

BTW - where is the `Job_ID` defined ? it should be a key in each object in `availability["Limits"]` 

```suggestion
    return CommandResults(
        outputs_prefix='ThreatZone.Limits',
        outputs_key_field='Job_ID',
        readable_output=readable_output,
        outputs=availability["Limits"]
    )
```

File: Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+
+
+def threatzone_sandbox_upload_sample(client: Client, args: Dict[str, Any]) -> CommandResults:
+    """Uploads the sample to the ThreatZone sandbox to analyse with required or optional selections."""
+    availability = client.threatzone_check_limits("sandbox")
+    if not availability["avaliable"]:
+        raise DemistoException(
+            f"Reason: {availability['Reason']}\nSuggestion: {availability['Suggestion']}\nLimits: {availability['Limits']}")
+
+    ispublic = args.get('isPublic')
```
Comment (ilappe) [inline Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py:376]: `isPublic` is not defined in the yml file 
and instead there is the arg `private` which are not been used in the code

File: Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+    readable_output = tableToMarkdown('SAMPLE UPLOADED', result)
+    uuid = result['uuid']
+    url = f"https://app.threat.zone/submission/{uuid}"
+    availability = client.threatzone_check_limits("sandbox")
+    return CommandResults(
+        readable_output=readable_output,
+        outputs_key_field='results',
+        outputs={'ThreatZone.Limits(val.Job_ID == obj.Job_ID)': availability["Limits"],
+                 'ThreatZone.Analysis(val.Job_ID == obj.Job_ID)': {'UUID': uuid, 'URL': url}}
+    )
```
Comment (ilappe) [inline Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py:419]: as mention above - please refactor the usage of CommandResults, and if you need to return two different keys to the context (like `ThreatZone.Limit` and `ThreatZone.Analysis`) you can return list of two CommandResults one for each of them

File: Packs/ThreatZone/Integrations/ThreatZone/README.md [hunk 4]
```diff
@@ -0,0 +1,791 @@
...
+#### Output
+
+
+| **Path**                 | **Type** | **Description**                                     |
+| -------------------------- | ---------- | ----------------------------------------------------- |
+| ThreatZone.Analysis.UUID | String   | UUID of the task created to analyze the submission. |
+| ThreatZone.Analysis.URL  | String   | URL of the task created to analyze the submission.  |
+
+### tz-get-report
+
```
Comment (ilappe) [inline Packs/ThreatZone/Integrations/ThreatZone/README.md:135]: this command are not defined in the yml file and not implemented in the py file

Please regenerate a README.md file.
To do that, please run the `demisto-sdk generate-docs` command (use the -e flag for using a commad_examples file) - for more information please see: https://xsoar.pan.dev/docs/concepts/demisto-sdk#generate-docs.

File: Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+        headers = {'Authorization': creds}
+
+        client = Client(
+            base_url=base_url,
+            verify=verify_certificate,
+            headers=headers,
+            proxy=proxy)
+
+        if command == 'test-module':
+            demisto.results(test_module(params))
```
Comment (ilappe) [inline Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py:524]:
```suggestion
            return_results(test_module(params))
```

**Review by ilappe — CHANGES_REQUESTED** [review]
Thank you @@vvelitkn-malwation for great contribution!

please see my comment 
and feel free to reach out to me with any questions - I'm available here or on slack :)

@melamedbn, will made the review of the palybooks etc.

**Comment by melamedbn** [comment]
Hi @vvelitkn-malwation,

I trust you're doing well. Firstly, thank you for your contribution; the effort you've put into the content is clear! I've taken some time to review it, and I'd like to share a few insights to help streamline and refine the playbooks:

### Playbooks: 
**Analyze File - Sandbox - ThreatZone, Analyze File - Static Scan - ThreatZone & Sanitize File - CDR - ThreatZone**

1. For consistency across all tasks, please maintain the same input logic. For instance, in task 1 and 8, it would be ideal to use `user inputs.File.EntryID` instead of just `File.EntryID`.
2. I recommend opting for the complex context method as opposed to the simple one. *(See attached screenshot for reference)*.

Screenshot 2023-10-16 at 10 22 42
**Simple**

Screenshot 2023-10-16 at 10 22 59
**Complex**

3. There appear to be some irregularities with the grid arrows that might benefit from correction.

image
**Current state**

image
**Preferred state**

4. **Recommendation:** If the primary difference between the playbooks is just the static/sandbox/CDR scan, consider consolidating the three playbooks into one. You could introduce a conditional task along with a playbook input, allowing users to select their preferred flow.

Your dedication is truly commendable, and these suggestions are geared towards enhancing the usability of your contribution. If you have any queries or if there's anything you'd like to discuss further, please don't hesitate to reach out.

Warm regards,
Ben

**Comment by vvelitkn-malwation** [comment]
> Hi @vvelitkn-malwation,
> 
> I trust you're doing well. Firstly, thank you for your contribution; the effort you've put into the content is clear! I've taken some time to review it, and I'd like to share a few insights to help streamline and refine the playbooks:
> ### Playbooks:
> 
> **Analyze File - Sandbox - ThreatZone, Analyze File - Static Scan - ThreatZone & Sanitize File - CDR - ThreatZone**
> 
>     1. For consistency across all tasks, please maintain the same input logic. For instance, in task 1 and 8, it would be ideal to use `user inputs.File.EntryID` instead of just `File.EntryID`.
> 
>     2. I recommend opting for the complex context method as opposed to the simple one. _(See attached screenshot for reference)_.
> 
> 
> Screenshot 2023-10-16 at 10 22 42 **Simple**
> 
> Screenshot 2023-10-16 at 10 22 59 **Complex**
> 
>     3. There appear to be some irregularities with the grid arrows that might benefit from correction.
> 
> 
> image **Current state**
> 
> image **Preferred state**
> 
>     4. **Recommendation:** If the primary difference between the playbooks is just the static/sandbox/CDR scan, consider consolidating the three playbooks into one. You could introduce a conditional task along with a playbook input, allowing users to select their preferred flow.
> 
> 
> Your dedication is truly commendable, and these suggestions are geared towards enhancing the usability of your contribution. If you have any queries or if there's anything you'd like to discuss further, please don't hesitate to reach out.
> 
> Warm regards, Ben

Hi Ben,
Thank you for your valuable feedback. I've incorporated an additional function for obtaining sanitized files. I will proceed with the necessary changes in playbooks and anticipate completion within the next few hours.

**Review by ilappe — CHANGES_REQUESTED** [review]
Thank @vvelitkn-malwation!

clean and nice implementation!
please see my few small comments

File: Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+    }
+
+    def create_res(readable_dict, output, exception=None):
+        if not exception:
+            readable_output = tableToMarkdown('Submission Result', readable_dict)
+        else:
+            readable_output = tableToMarkdown('Submission Result', {"Exception": str(exception)})
+        return [
+            CommandResults(
+                outputs_prefix='ThreatZone.Analysis',
```
[the remark this answers is not in the document]
Comment (vvelitkn-malwation) [reply Packs/ThreatZone/Integrations/ThreatZone/ThreatZone.py:289]: ThreatZone.Result is deprecated, we use ThreatZone.Analysis instead, fixed in YML.

**Comment by melamedbn** [comment]
Hi @vvelitkn-malwation,

Did you get a chance to address my feedback?

Thanks,
Ben

**Comment by vvelitkn-malwation** [comment]
> Hi @vvelitkn-malwation,
> 
> Did you get a chance to address my feedback?
> 
> Thanks, Ben

Yes, I done changes according to your valuable feedbacks.

**Review by melamedbn — APPROVED** [review]

**Review by ilappe — APPROVED** [review]
veri nice!
the only one comment
i think it will be better to define unique context path for the 3 upload types
`static` `cdr` and `sandbox`
as currently the context output are ThreatZone.Submission for all of them
and if customer will want to know which of the outputs url are for the `static` and which of the `cdr `- he can’t
so may be the good context output should be `ThreatZone.CdrSubmission`  `ThreatZone.StaticSubmission`  etc
but if you think this is not so important - we can go forward in the process

**Pull request merged**